cogeo_mosaic/backend/base.py

```diff
@@ -0,0 +1,70 @@
+"""cogeo_mosaic.backend.base: base Backend class."""
+
+import abc
+from contextlib import AbstractContextManager
+from typing import Dict, Optional
+
+from cogeo_mosaic import version as mosaic_version
+from cogeo_mosaic.backend.utils import get_hash
+from cogeo_mosaic.utils import create_mosaic
+
+
+class BaseBackend(AbstractContextManager):
+    mosaic_def: Optional[Dict]
+
+    @abc.abstractmethod
+    def __init__(self, *args, **kwargs):
+        """Connect to backend"""
+
+    def __enter__(self):
+        """Support using with Context Managers"""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers"""
+
+    def metadata(self) -> Dict:
+        """Retrieve Mosaic metadata
+
+        Returns
+        -------
+        MosaicJSON as dict without `tiles` key.
+        """
+        if self.mosaic_def is None:
+            return {}
+
+        return {k: v for k, v in self.mosaic_def.items() if k != "tiles"}
+
+    @abc.abstractmethod
+    def tile(self, x: int, y: int, z: int):
+        """Retrieve assets for tile."""
+
+    @abc.abstractmethod
+    def point(self, lng: float, lat: float):
+        """Retrieve assets for point."""
+
+    def create(self, *args, **kwargs):
+        """Create new MosaicJSON and upload to backend."""
+
+        self.mosaic_def = create_mosaic(*args, **kwargs)
+        self.upload(self.mosaic_def)
+
+    @property
+    def mosaicid(self) -> str:
+        return get_hash(body=self.mosaic_def, version=mosaic_version)
+
+    @property
+    def quadkey_zoom(self) -> Optional[int]:
+        if self.mosaic_def:
+            return self.mosaic_def.get("quadkey_zoom", self.mosaic_def["minzoom"])
+
+        return None
+
+    @abc.abstractmethod
+    def upload(self, mosaic: Dict):
+        """Upload new MosaicJSON to backend."""
+
+    # Commented to allow child classes to not implement this
+    # @abc.abstractmethod
+    def update(self, mosaic: Dict):
```

Comment:
Give mosaicJSON is a well defined spec, I think we should add some stricter validation than standard python types.  I'm a big fan of `pydantic` which will throw validation errors if the mosaicJSON doesn't match the defined model.  This will make it easier to add new backends because the mosaicJSON definition is explicitly defined in the code and prevent exceptions (ex `KeyError`) from happening downstream.

```python
from pydantic import BaseModel

class MosaicJSON(BaseModel):
    """https://github.com/developmentseed/mosaicjson-spec/tree/master/0.0.2"""

    minzoom: int
    maxzoom: int
    tiles: Dict[str, List[str]]
    bounds: List[NumType] = [-180, -90, 180, 90]
    name: Optional[str] = "null"
    description: Optional[str] = "null"
    attribution: Optional[str] = "null"
    center: Optional[List[NumType]] = "null"
    version: Optional[str] = "1.0.0"
    quadkey_zoom: Optional[int] = 0
    mosaicjson: Optional[str] = "0.0.2"
```

Reply:
Ooh does this mean once you define your code as listed you can just pass
```
def update(self, mosaic: MosaicJSON):
```
? That's really cool. Although I don't think `mypy` is currently actually checking types, though @vincentsarago can correct me if I'm wrong.

Reply:
I love this idea. 
here the version of the model I had on my side.

```python
class MosaicJSON(BaseModel):
    """
    MosaicJSON model.

    Based on https://github.com/developmentseed/mosaicjson-spec

    """

    mosaicjson: str = "0.0.2"
    name: Optional[str]
    description: Optional[str]
    version: str = "1.0.0"
    attribution: Optional[str]
    minzoom: int = Field(0, ge=0, le=30)
    maxzoom: int = Field(30, ge=0, le=30)
    quadkey_zoom: Optional[int]
    bounds: List[float] = [-180, -90, 180, 90]
    center: Optional[Tuple[float, float, int]]
    tiler: Dict[str, List[str]]
```

Reply:
Do we require that `tiles` exist? For example, loading the Mosaic definition from DynamoDB doesn't include the `tiles` key. So to be consistent, we could either allow `tiles` to be `Optional[Dict[str, List[str]]]`, or when loading the mosaic definition from dynamodb add an empty `dict`

Reply:
For now, I just set `tiles` to be an empty `dict` when loading from DynamoDB

Reply:
I also have `self.mosaic_def` stored internally as an instance of the pydantic class, instead of a plain `dict`

Reply:
@kylebarron the the `tiles` I think it's before to raise an exception, the `tiles` is theoretically set but just not accessible with dynamoDB

Reply:
I don't know what you meant?

Reply:
I mean `tiles` shouldn't return empty because in theory `tiles` is present in the mosaicjson document, it's just that dynamoDB backend is not able to return the whole `tiles` array. That's why I think this should raise an exception in this backend